File: src/tracker.cpp
#include "voice_detection/tracker.hpp"

#include "voice_detection/math_utils.hpp"

#include <algorithm>
#include <limits>

namespace voice_detection
{

SourceTracker::SourceTracker(float max_angle_delta_deg, std::int64_t max_age_ms)
: max_angle_delta_deg_(max_angle_delta_deg), max_age_ms_(max_age_ms)
{
}

std::string SourceTracker::assign(std::int64_t stamp_ms, const std::optional<float> & azimuth_deg)
{
  expire(stamp_ms);
  if (!azimuth_deg.has_value()) {
    return "voice_mono";
  }

  const float azimuth = normalizeAngleDeg(*azimuth_deg);
  TrackState * best = nullptr;
  float best_delta = std::numeric_limits<float>::max();
  for (auto & track : tracks_) {
    if (!track.azimuth_deg.has_value()) {
      continue;
    }
    const float delta = angularDistanceDeg(azimuth, *track.azimuth_deg);
    if (delta < best_delta) {
      best_delta = delta;
      best = &track;
    }
  }
  if (best != nullptr && best_delta <= max_angle_delta_deg_) {
    best->azimuth_deg = azimuth;
    best->last_seen_ms = stamp_ms;
    return best->track_id;
  }

  TrackState track;
  track.track_id = "voice_" + std::to_string(next_id_++);
  track.azimuth_deg = azimuth;
  track.last_seen_ms = stamp_ms;
  tracks_.push_back(track);
  return track.track_id;
}

std::vector<std::string> SourceTracker::trackedIds(std::int64_t stamp_ms)
{
  expire(stamp_ms);
  std::vector<std::string> ids;
  ids.reserve(tracks_.size());
  for (const auto & track : tracks_) {
    ids.push_back(track.track_id);
  }
  return ids;
}

void SourceTracker::expire(std::int64_t stamp_ms)
{
  tracks_.erase(
    std::remove_if(
      tracks_.begin(), tracks_.end(),
      [this, stamp_ms](const TrackState & track) {
        return stamp_ms - track.last_seen_ms > max_age_ms_;
      }),
    tracks_.end());
}

}  // namespace voice_detection

File: voice_detection/tracker.py
from __future__ import annotations

from dataclasses import dataclass

from .doa import normalize_angle_deg


@dataclass
class _TrackState:
    track_id: str
    azimuth_deg: float | None
    last_seen_ms: int


class SourceTracker:
    def __init__(self, max_angle_delta_deg: float = 25.0, max_age_ms: int = 1200):
        self.max_angle_delta_deg = max_angle_delta_deg
        self.max_age_ms = max_age_ms
        self._tracks: list[_TrackState] = []
        self._next_id = 1

    def assign(self, stamp_ms: int, azimuth_deg: float | None) -> str:
        self._expire(stamp_ms)
        if azimuth_deg is None:
            return "voice_mono"
        azimuth = normalize_angle_deg(azimuth_deg)
        best: _TrackState | None = None
        best_delta = float("inf")
        for track in self._tracks:
            if track.azimuth_deg is None:
                continue
            delta = angular_distance_deg(azimuth, track.azimuth_deg)
            if delta < best_delta:
                best = track
                best_delta = delta
        if best is not None and best_delta <= self.max_angle_delta_deg:
            best.azimuth_deg = azimuth
            best.last_seen_ms = stamp_ms
            return best.track_id
        track = _TrackState(track_id=f"voice_{self._next_id}", azimuth_deg=azimuth, last_seen_ms=stamp_ms)
        self._next_id += 1
        self._tracks.append(track)
        return track.track_id

    def tracked_ids(self, stamp_ms: int) -> tuple[str, ...]:
        self._expire(stamp_ms)
        return tuple(track.track_id for track in self._tracks)

    def _expire(self, stamp_ms: int) -> None:
        self._tracks = [
            track for track in self._tracks
            if stamp_ms - track.last_seen_ms <= self.max_age_ms
        ]


def angular_distance_deg(a: float, b: float) -> float:
    delta = abs((a - b + 180.0) % 360.0 - 180.0)
    return min(delta, 360.0 - delta)

